### ISSUE TITLE
Is the z/OS REIPL feature implemented in Hercules?

### ISSUE THREAD

**Issue by Vf58** [SDL-Hercules-390/hyperion#282]
Hello,

Adding the following statement to ` SYS1.PARMLIB(DIAGxx)` member:

```
AUTOIPL SADMP(NONE) MVS(F310,F32123M)
```

is supposed to allow you to perform a re-ipl of the sytem via the command:

```
V XCF,HERC,OFFLINE,,REIPL
```

But when I try that, I receive:

```
IGV010I IN PARMLIB MEMBER=DIAG00 ON LINE 4:
AUTOIPL WAS NOT PROCESSED, DUE TO AN ENVIRONMENT ERROR
```

Is there some z/Architecture Facility (or Hercules facility?) that I need to activate/enable beforehand? Or is support for this feature simply not implemented yet in Hercules?

Just so you know, if it's not implemented yet, that's okay. It's not a big deal. I can still more or less accomplish the same thing via Hercules's HAO feature.

Thanks in advance for any help/enlightenment you can provide on the subject.

Mich

**Comment by Fish-Git** [comment]
_**STANDARD DISCLAIMER:** I don't know z/OS. I have almost **no practical experience** to speak of. My background is mostly DOS/VSE and VM/SP. I've dabbled in z/OS a little bit over the years only as a result of my having worked in Hercules for so long, but practically speaking, my z/OS skills are virtually non-existent. Others know z/OS much better than me. I might know a teeny tiny bit, but it's not very much, so please keep that in mind when you discuss z/OS things with me. Don't presume I know what you're talking about!_
___

Hi Mich!

I'm _guessing_ that this is being caused by our lack of proper DIAGNOSE X'308' support, which I suspect is what z/OS might be trying to use? We do have "some" DIAGNOSE X'308' support, but it is _severely_ lacking and known to not be right.

I also notice that when the function code is 3 or 4, Hercules logs message `HHC01900I Diagnose 0x308 called: System is re-ipled`. Do you see that anywhere in your Hercules log? If so, then the re-ipl _should_ have worked. The stand-alone dump part of the request might not have worked (that's part of the code we're lacking), but the re-ipl part _should_ have worked from what I can see.

If the code was 5 or 6 however, then it definitely isn't going to work. Not only do we not have the proper code in place to process the request, the code we do have appears to be wrong. It's setting GPR 1 to the value 1 (indicating "success") instead of setting the r1+1 register to that value.

So... support for codes 5 and 6 are non-existent and broken, and support for code 4 is incomplete (it's missing the sadump logic). Code 3 however (plain re-IPL) _should_ work. Codes 4, 5 and 6? No. Code 3? Yes.

Which DIAG X'308' request code z/OS is issuing I don't know and don't know how to find out. You could do an instruction trace, but that will likely take forever and create a humongous log file most of which will be meaningless.

I'll have to add some debugging code to our DIAG X'308' logic to help see what's going on now that I know how to recreate the problem (which I have not tried doing yet).

Thanks for reporting this, Mich.

We will _try_ to get it fixed (or at least try to get it to work a little better than it currently is), but can of course make no guarantees. It might take a while.

I suggest sticking with the "HAO" technique for now. &nbsp; `:)`
___
_**STANDARD DISCLAIMER:** I don't know z/OS. I have almost **no practical experience** to speak of. My background is mostly DOS/VSE and VM/SP. I've dabbled in z/OS a little bit over the years only as a result of my having worked in Hercules for so long, but practically speaking, my z/OS skills are virtually non-existent. Others know z/OS much better than me. I might know a teeny tiny bit, but it's not very much, so please keep that in mind when you discuss z/OS things with me. Don't presume I know what you're talking about!_

**Comment by Fish-Git** [comment]
@Vf58

Additional Information:

I spent a few minutes trying to recreate this problem, but didn't get very far. &nbsp; `:(`

The first thing I did was to issue the `DISPLAY DIAG` command on the master console, which showed that `AUTOIPL SADMP(NONE) MVS(NONE)` was active.

So I then edited my my `ADCD.Z23A.PARMLIB(DIAG00)` member and added:

```
AUTOIPL SADMP(NONE) MVS(0A80,0A82WSM)
```

followed by the master console command `SET DIAG=00` to activate it. This however resulted in messages:

```
IGV010I IN PARMLIB MEMBER=DIAG00 ON LINE 30:
AUTOIPL WAS NOT PROCESSED, DUE TO AN ENVIRONMENT ERROR
```

(My `DIAG00` member has a bunch of comments in it, so line 30 is my `AUTOIPL` statement.)

I then tried specifying `LAST` for the `MVS` parameter instead (instead of a specific device and loadparm), a la:

```
AUTOIPL SADMP(NONE) MVS(LAST)
```

but the same error occurred. &nbsp; `:(`

And each time, after issuing the `SET DIAG=00` command (which would alway fail), I followed it with a `DISPLAY DIAG` command to verify the current settings, and sure enough, `AUTOIPL` was still set to `SADMP(NONE) MVS(NONE)` (i.e. no change was made; the set diag command indeed failed and did not take effect).

I also set a tracepoint in the diagnose function to trace all diagnose calls, and the only diagnose calls that z/OS 2.3 appears to make are X'204' and X'9C'. That is to say, DIAG X'308' seems to never be called (which I guess makes sense if z/OS knows it's not going to work).

I have _not_ tried tracing all of the diagnose calls that are made during IPL however. It could be that z/OS checks to see what support is available during NIP (Nucleus Initialization Phase?) and maybe saves this information somewhere in some internal table maybe?

Or maybe the problem (bug) is in our DIAG X'204' (LPAR) logic somewhere? Maybe we're not returning the correct information or not supporting a specific call/request type?

Bottom line: it doesn't look like this issue (this problem/"bug") is going to get fixed any time soon, if ever. &nbsp; `:(`

If someone else (some other Hercules developer/hacker) wants to look into this issue a little deeper, I'd appreciate it. But until then, I'm just going to have to mark this one as "Won't Fix" since it appears fixing it is going to take more effort than it's probably worth. I don't believe this particular lack of support in Hercules is causing any serious problems to anyone, given that there _are_ viable workarounds that are available (such as the HAO technique that Mich mentioned himself).

So I'm going to close this issue as "Won't Fix". If someone wants to re-open it and work on it, please feel free to do so!

(Sorry Mich!) &nbsp; `:(`

**Comment by Vf58** [comment]
Hi Fish,


Thank you for your answer. And don't be sorry, we both agree we can easily use HAO.
Just for memory, but you certainly know it, the same functionality exist on Z/VM with "shutdown ... reipl ... iplparms ...'. Does it use the same mechanism? I presume, but I really don't know.

(https://www.ibm.com/support/knowledgecenter/en/SSB27U_6.4.0/com.ibm.zvm.v640.hcpb7/shtdown.htm)

Thanks again, :-)

Mich.

**Comment by Fish-Git** [comment]
> Just for memory, but you certainly know it, the same functionality exist on Z/VM with "shutdown ... reipl ... iplparms ...". Does it use the same mechanism? I presume, but I really don't know.

The z/VM shutdown reipl command works fine, BUT... I have not tried it with the "iplparms" option. I've only ever used the plain "shutdown system xxx reipl" command, and as I said, it works fine.

Does it use the same mechanism as z/OS? I don't know! But I'm guessing that it doesn't. At least not for a plain reipl. For a plain reipl, it appears it is doing the reipl itself by issuing the I/O to the sysres device to kick off a normal boot process.

For a reipl with the iplparms option? Maybe it behaves differently. I don't know. I've never tried it. My z/VM system doesn't use/need any special iplparms so I don't know what iplparms I should use as a test. Have _you_ tried it? Does it work for you?

**Comment by moshix** [comment]
Reipl with and without iplparms works fine for me with current SDL and z/VM 6.2. Haven’t tested with earlier or later versions

**Comment by Vf58** [comment]
Hi,

Thank you for the test. I cannot try with Z/vm, I do not have one installed. I was VM & Dos/vse system engineer till 1993 when I start doing the same job on MVS and I'm afraid I loose 99.98% of my skills ;-) 

Good to know it is working on Z/VM.
I could imagine the way the order is given to perform the IPL  is using the same principle (why should it be different? Why not! ;-) ). The difference could be in the Z/OS command which also affect the SYSPLEX even in monoplex.
I'm also sure it is not using BCPII....

In all cases it is not a problem, I'll use HAO.

Thank you again

Mich
